apis/placement/v1alpah1/override_types.go

```diff
@@ -0,0 +1,186 @@
+	// When there are two rules selecting the same fields on the target cluster, the last one will win.
+	// You can have 1-100 rules.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=100
+	// +required
+	OverrideRules []OverrideRule `json:"overrideRules"`
+}
+
+type OverrideRule struct {
+	// ClusterSelectors selects the target clusters.
+	// The resources will be overridden before applying to the matching clusters.
+	// If ClusterSelector is not set, it means selecting ALL the member clusters.
+	// +optional
+	ClusterSelector *placementv1beta1.ClusterSelector `json:"clusterSelector,omitempty"`
+
+	// JSONPatchOverrides defines a list of JSON patch override rules.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=20
+	// +required
+	JSONPatchOverrides []JSONPatchOverride `json:"jsonPatchOverrides"`
+}
+
+// +genclient
+// +genclient:Namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceOverride defines a group of override policies about how to override the selected namespaced scope resources
+// to target clusters.
+type ResourceOverride struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of ResourceOverrideSpec.
+	// +required
+	Spec ResourceOverrideSpec `json:"spec"`
+}
+
+// ResourceOverrideSpec defines the desired state of the Override.
+type ResourceOverrideSpec struct {
+	// ResourceSelectors is an array of selectors used to select namespace scoped resources. The selectors are `ORed`.
+	// You can have 1-20 selectors.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=20
+	// +required
+	ResourceSelectors []ResourceSelector `json:"resourceSelectors"`
+
+	// Policy defines how to override the selected resources on the target clusters.
+	// +required
+	Policy *OverridePolicy `json:"policy"`
+}
+
+// ResourceSelector is used to select namespace scoped resources as the target resources to be placed.
+// All the fields are `ANDed`. In other words, a resource must match all the fields to be selected.
+type ResourceSelector struct {
+	// Group name of the namespace-scoped resource.
+	// Use an empty string to select resources under the core API group (e.g., services).
+	// +required
+	Group string `json:"group"`
+
+	// Version of the namespace-scoped resource.
+	// +required
+	Version string `json:"version"`
+
+	// Kind of the namespace-scoped resource.
+	// +required
+	Kind string `json:"kind"`
+
+	// Namespace of the namespace-scoped resource.
+	// +optional
+	// Default is empty, which means inherit from the parent object scope.
+	Namespace string `json:"namespace,omitempty"`
```

Comment:
the resourceOverride is already a namespaced object, we don't need a namespace field here. Aka, we should not allow a resourceOverride object in namespace A says that it will override some object in namespace B.

Reply:
SG. The namespace is redundant. Will remove.